r/src/RecordBatch.cpp

```diff
@@ -63,17 +63,26 @@ List RecordBatch__to_dataframe(const std::shared_ptr<arrow::RecordBatch>& batch)
 }
 
 // [[Rcpp::export]]
-std::shared_ptr<arrow::RecordBatch> read_record_batch_(std::string path) {
-  std::shared_ptr<arrow::io::ReadableFile> stream;
+std::shared_ptr<arrow::RecordBatch> read_record_batch_RandomAccessFile(
+    const std::shared_ptr<arrow::io::RandomAccessFile>& stream) {
   std::shared_ptr<arrow::ipc::RecordBatchFileReader> rbf_reader;
-
-  R_ERROR_NOT_OK(arrow::io::ReadableFile::Open(path, &stream));
   R_ERROR_NOT_OK(arrow::ipc::RecordBatchFileReader::Open(stream, &rbf_reader));
 
   std::shared_ptr<arrow::RecordBatch> batch;
   R_ERROR_NOT_OK(rbf_reader->ReadRecordBatch(0, &batch));
 
-  R_ERROR_NOT_OK(stream->Close());
+  return batch;
+}
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::RecordBatch> read_record_batch_BufferReader(
```

Comment:
This function name does not suggest that it reads a stream (including the schema). The Python version of this does not read a stream

Reply:
It reads a record batch from a BufferReader. 

But yeah this is likely to change once more C++ classes can be held on the r side. 

Should it just read the record batch and not the schema ?

Reply:
We need to go carefully through all the modes of operation when it comes to IPC, as listed in https://issues.apache.org/jira/browse/ARROW-3498. We have support for all modes in Python at the moment:

* Reading a stream
* Reading a file
* Reading a schema alone
* Reading a single record batch alone given a known schema